dds/src/implementation/rtps/types.rs

```diff
@@ -345,47 +320,63 @@ pub const PROTOCOLVERSION_2_2: ProtocolVersion = ProtocolVersion { major: 2, min
 pub const PROTOCOLVERSION_2_3: ProtocolVersion = ProtocolVersion { major: 2, minor: 3 };
 pub const PROTOCOLVERSION_2_4: ProtocolVersion = ProtocolVersion { major: 2, minor: 4 };
 
-impl From<ProtocolVersion> for [u8; 2] {
-    fn from(value: ProtocolVersion) -> Self {
-        [value.major, value.minor]
+impl ProtocolVersion {
+    pub fn new(major: u8, minor: u8) -> Self {
+        Self { major, minor }
     }
-}
-impl From<[u8; 2]> for ProtocolVersion {
-    fn from(value: [u8; 2]) -> Self {
-        Self {
-            major: value[0],
-            minor: value[1],
-        }
+    pub fn major(&self) -> u8 {
+        self.major
+    }
+    pub fn minor(&self) -> u8 {
+        self.minor
     }
 }
 
 /// VendorId_t
 /// Type used to represent the vendor of the service implementing the RTPS protocol. The possible values for the vendorId are assigned by the OMG.
 /// The following values are reserved by the protocol: VENDORID_UNKNOWN
-pub type VendorId = [u8; 2];
-pub const VENDOR_ID_UNKNOWN: VendorId = [0, 0];
-pub const VENDOR_ID_S2E: VendorId = [99, 99];
+#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct VendorId([u8; 2]);
+pub const VENDOR_ID_UNKNOWN: VendorId = VendorId([0, 0]);
+pub const VENDOR_ID_S2E: VendorId = VendorId([99, 99]);
+
+impl VendorId {
+    pub fn new(value: [u8; 2]) -> Self {
+        Self(value)
+    }
+}
 
+impl AsRef<[u8; 2]> for VendorId {
+    fn as_ref(&self) -> &[u8; 2] {
+        &self.0
+    }
+}
 /// Count_t
 /// Type used to hold a count that is incremented monotonically, used to identify message duplicates.
 #[derive(Clone, Copy, PartialEq, Eq, Debug, serde::Serialize, serde::Deserialize)]
-pub struct Count(pub i32);
+pub struct Count(i32);
 
+impl Count {
+    pub const fn new(value: i32) -> Self {
+        Self(value)
+    }
+    pub const fn wrapping_add(self, rhs: i32) -> Self {
+        Self(self.0.wrapping_add(rhs))
+    }
+}
+impl PartialOrd<Count> for Count {
+    fn partial_cmp(&self, other: &Count) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
 impl AddAssign for Count {
     fn add_assign(&mut self, rhs: Self) {
         self.0 += rhs.0;
     }
 }
-
-impl From<i32> for Count {

```

Comment:
Is there any rule/guideline to decide between new and From trait implementation?

Reply:
It's implicit described in https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv
The intent is though the important thing here: The new()'s intent is to construct a type while from()'s intent is to convert a type. Since I want to suggest that the type needs to be constructed from bytes, it looks in the usage more explicit during it's construction. 
Though I would like to remove the new()'s and From() impl's altogether and use only the mapping. But that requires some more re-organization.